api/client/net_test.go

```diff
@@ -0,0 +1,66 @@
+package client
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	n "github.com/textileio/powergate/net"
+	"github.com/textileio/powergate/net/rpc"
+)
+
+func TestNet(t *testing.T) {
+	c, done := setupNet(t)
+	defer done()
```

Comment:
Not sure if the is the best setup, but other setups brought difficulties. This setup has one top level test func that creates the server and client and then all nested `t.Run` calls use those.

The other options I explored were:

- Creating stand alone tests that each created their own server and client. In this case, I was getting errors about the Prometheus port already being in use, presumably because each stand alone test was running with some amount of parallelism. I'd also imagine this would be quite heavy on the CI machine.
- Using `TestMain` to setup a single server and client shared between all stand alone tests. I actually really like this idea, but all of our code for setting up the server depends on a `testing.T` being passed in (which doesn't exist in `TestMain`). So this is possible, but required refactoring of a bunch of test setup code for creating the server.

Open to input here for sure.

Reply:
I think it's better that each subtest runs a clean environment. The thing with sharing it is that one test might run correctly because other subtests leave a _convenient_ state, and test order might include other hidden magics.

Closing the server should liberate all resources that were created... that includes all files, ports, etc. So that might be an indication that some closing thing isn't closing. I you want I can take a look.